server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportEnsureClusterStateVersionAppliedAction.java

```diff
@@ -0,0 +1,191 @@
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.AbstractTransportRequest;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+
+/**
+ * An action that waits for a given cluster state version to be applied on provided set of nodes in the cluster.
+ */
+public class TransportEnsureClusterStateVersionAppliedAction extends TransportNodesAction<
+    EnsureClusterStateVersionAppliedRequest,
+    EnsureClusterStateVersionAppliedResponse,
+    TransportEnsureClusterStateVersionAppliedAction.NodeRequest,
+    TransportEnsureClusterStateVersionAppliedAction.NodeResponse,
+    Void> {
+    public static final ActionType<EnsureClusterStateVersionAppliedResponse> TYPE = new ActionType<>(
+        "internal:cluster/nodes/state/ensure_version"
+    );
+
+    private static final Logger logger = LogManager.getLogger(TransportEnsureClusterStateVersionAppliedAction.class);
+
+    private final ThreadPool threadPool;
+
+    @Inject
+    public TransportEnsureClusterStateVersionAppliedAction(
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        ThreadPool threadPool
+    ) {
+        super(
+            TYPE.name(),
+            clusterService,
+            transportService,
+            actionFilters,
+            NodeRequest::new,
+            threadPool.executor(ThreadPool.Names.GENERIC)
+        );
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    protected EnsureClusterStateVersionAppliedResponse newResponse(
+        EnsureClusterStateVersionAppliedRequest request,
+        List<NodeResponse> nodeResponses,
+        List<FailedNodeException> failures
+    ) {
+        return new EnsureClusterStateVersionAppliedResponse(clusterService.getClusterName(), nodeResponses, failures);
+    }
+
+    @Override
+    protected NodeRequest newNodeRequest(EnsureClusterStateVersionAppliedRequest request) {
+        return new NodeRequest(request.clusterStateVersion(), request.nodeTimeout());
+    }
+
+    @Override
+    protected NodeResponse newNodeResponse(StreamInput in, DiscoveryNode node) throws IOException {
+        return new NodeResponse(in, node);
+    }
+
+    @Override
+    protected NodeResponse nodeOperation(NodeRequest request, Task task) {
+        /// We are using [#nodeOperationAsync].
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected void nodeOperationAsync(NodeRequest request, Task task, ActionListener<NodeResponse> listener) {
+        var cancellableTask = (CancellableTask) task;
+
+        Predicate<ClusterState> predicate = (ClusterState state) -> cancellableTask.isCancelled()
+            || state.version() == request.clusterStateVersion;
```

Comment:
Hmm, does this need to be `>=`? And do we need a test case for that?

Reply:
Yes, and yes :)